include/rmm/mr/device/detail/arena.hpp

```diff
@@ -0,0 +1,616 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/detail/aligned.hpp>
+#include <rmm/detail/error.hpp>
+
+#include <cuda_runtime_api.h>
+
+#include <algorithm>
+#include <limits>
+#include <mutex>
+#include <set>
+#include <unordered_map>
+
+namespace rmm {
+namespace mr {
+namespace detail {
+namespace arena {
+
+constexpr std::size_t superblock_size = 1u << 26u;  ///< Size of a superblock (64 MiB).
+
+/**
+ * @brief Represents a chunk of memory that can be allocated and deallocated.
+ *
+ * A fixed-sized block obtained from the global arena is called a "superblock". Only a superblock
+ * can be returned to the global arena.
+ */
+struct block {
+  void* pointer;     ///< Raw memory pointer.
+  std::size_t size;  ///< Size in bytes.
+  bool is_head;      ///< Indicates whether pointer was allocated from upstream.
```

Comment:
I thought I did this in a previous review, but currently any time you're returning a default constructed `block` it is returning an uninitialized object. Using any of the member functions like `is_valid` is then undefined behavior. 
```suggestion
  void* pointer{};     ///< Raw memory pointer.
  std::size_t size{};  ///< Size in bytes.
  bool is_head{};      ///< Indicates whether pointer was allocated from upstream.
```

Reply:
Hmm, I don't think that's true. Zero initialization is performed when no constructor is defined (https://en.cppreference.com/w/cpp/language/zero_initialization).
```
$ cat foo.cpp
```
```cpp
#include <iostream>

struct block {
  void* pointer;
  std::size_t size;
  bool is_head;
};

int main()
{
  block b{};
  std::cout << b.pointer << ", " << b.size << ", " << b.is_head << "\n";
}
```
```
$ g++ foo.cpp
$ ./a.out
0, 0, 0
```

Reply:
That's only true if you brace init everywhere. It wouldn't be true if you had:
```
block b;
std::cout << b.pointer << ", " << b.size << ", " << b.is_head << "\n";
```

There's no additional overhead to the in class initialization and it makes it safer overall.

Reply:
Done.